db/commitlog/commitlog.cc

```diff
@@ -1899,6 +1902,18 @@ void db::commitlog::segment_manager::discard_completed_segments(const cf_id_type
     discard_unused_segments();
 }
 
+future<> db::commitlog::segment_manager::force_new_active_segment() noexcept {
+    if (_segments.empty() || !_segments.back()->is_still_allocating()) {
+        co_return;
+    }
+
+    auto& s = _segments.back();
+    if (s->position()) { // check used.
+        co_await s->close();
+        discard_unused_segments();
+    }
+}
+
```

Comment:
Q: let's say we're forcing a flush due to the hitting the hard limit. Will we deadlock? Or will just new commits wait until the flush completes (it's sad since it's a long wait, but okay since it's an edge case).

The fact that commitlog depends on activities that need commitlog is sad. Maybe we should direct all those internal table writes to the schema commitlog.

Reply:
This will not, no, since the salient point is a segment exists and it has data. It might cause some really inefficient processing though, with writes + flushes of very small segments.

Reply:
I'm talking about this dependency loop:
1. commitlog is at the hard limit, so it wants to flush
2. flush now needs to allocate a new segment
3. we can't allocate a new segment because we're at the hard limit

Reply:
We don't actually allocate a new segment here. We just discard any existing one. Allocation will not happen until a new mutation comes in, which might block, yes, but this memtable flush/compaction will still be able to finish and release segments.